### PR TITLE
[4] Introducing Joomla’s first blank state

### PR DESCRIPTION
**Pull request by PhilETaylor**
### Summary of Changes

So [I mentioned about the possibility of blank states](https://github.com/joomla/joomla-cms/issues/30887#issuecomment-702959686) in Joomla 4 as a comment on someone's PR a long time ago. It received quite a bit of positive feedback but no one has actually attempted it

So here is "an attempt"

The concept is that rather than a new user getting a first time out of the box experience with "no matching results" or a state that is unhelpful, that the user receives a welcome and an introduction to each area. 

This PR is a PROOF OF CONCEPT for Articles ONLY - if it received praise then I might do the others, and others can join in too..

The blank state template is ONLY USED when you have no articles in your database. It is NOT used if a search/filter produces no results (in that case, you would correctly get "No Matching Results" message like you currently do. 

### To be clear

This is not a new concept. Everyone else is already doing it. 

More examples of Empty State Design:
https://www.toptal.com/designers/ux/empty-state-ux-design

### Testing Instructions

Apply PR on a site with no articles. 
Add article - blank state design disappears and you see a list of articles
filter the articles by something (search for kashfdksjhgfkjsdhf) and see that you DONT see the blank slate but you DO get "No Matching Results"
Trash your article -  and see that you DONT see the blank slate
Empty your trash (leaving ZERO articles in the db)  and see that you DO now see the blank slate

### Actual result BEFORE applying this Pull Request

At the moment the out of the box experience (without demo data) is a bunch of left menu items that all lead to blank pages saying "no matching results" 

User is "welcomed" to Joomla with a negative messages

Screenshot 2021-04-23 at 14 27 04

### Expected result AFTER applying this Pull Request

Empty State Design / Blank state Design pattern is used. 

### Documentation Changes Required

Screenshots. 

Screenshot 2021-04-23 at 15 15 55

### CODE REVIEW

**Comment by PhilETaylor** [comment]
Mobile: 

Screenshot 2021-04-23 at 14 22 25

**Comment by richard67** [comment]
@PhilETaylor Please remove any emojis from title and description of PRs. They can confuse our CI tools.

**Comment by richard67** [comment]
... and the issue tracker can be confused, too.

**Comment by PhilETaylor** [comment]
>@PhilETaylor Please remove any emojis from title and description of PRs. They can confuse our CI tools. 

/facepalm /voidOfExpression :-)

**Comment by richard67** [comment]
Beside the above comment (which is fixed now), I like this PR. Much more user friendly way to show there are no items.

**Comment by brianteeman** [comment]
1. As a non-coder I didn't understand what you meant by blankstate. Now I understand it to mean "no records yet" ?
2. Can you add aria-hidden="true" to the icon please
3. The voice you have used for the message is very informal, which is fine, but everywhere else we use a more formal voice such as "No articles have been created yet,"

**Comment by PhilETaylor** [comment]
1. Blank State = Out of the box experience = The Empty State
2. yup 
3. I fully expected the text to be completely changed by you and others before it was merged. It was copied from Digital Ocean.

Screenshot 2021-04-23 at 14 39 55

Also the "Learn More" button goes to docs.joomla.org - but to an old page... might there be somewhere better... 

Also how much help is enough, for example, the above screenshot shows more help than just one button etc...

**Comment by brianteeman** [comment]
Actually I can see now that I test it that the text you propose is not correct as it is displayed when you have created articles they just are trashed.

And if you have trashed all the articles then there is no way to select them to permanently delete or to restore them (unless you create a new article). So I guess the check for "blankstate" is not correct.

**Comment by PhilETaylor** [comment]
> Actually I can see now that I test it that the text you propose is not correct as it is displayed when you have created articles they just are trashed.

No that should never happen. It was specifically tested for during design. 

If you had 1million articles in the trash and zero in the filtered results, then you should never see the blank state.

File: administrator/components/com_content/tmpl/articles/blankstate.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	<div class="px-4 py-5 my-5 text-center">
+		<span aria-hidden="true" class="fa-8x icon-copy mb-4 article"></span>
+		<h1 class="display-5 fw-bold"><?php echo Text::_('COM_CONTENT_BLANKSTATE_TITLE'); ?></h1>
+		<div class="col-lg-6 mx-auto">
+			<p class="lead mb-4">
+				<?php echo Text::_('COM_CONTENT_BLANKSTATE_CONTENT'); ?>
+			</p>
+			<div class="d-grid gap-2 d-sm-flex justify-content-sm-center">
+				<a href="<?php echo Route::_('index.php?option=com_content&view=article&layout=edit'); ?>" class="btn btn-primary btn-lg px-4 me-sm-3"><?php echo Text::_('COM_CONTENT_BLANKSTATE_BUTTON_ADD'); ?></a>
+				<a href="https://docs.joomla.org/Adding_a_new_article" class="btn btn-outline-secondary btn-lg px-4"><?php echo Text::_('COM_CONTENT_BLANKSTATE_BUTTON_LEARNMORE'); ?></a>
```
Comment (brianteeman) [inline administrator/components/com_content/tmpl/articles/blankstate.php]: To make the link work with all languages it should be changed

```suggestion
				<a href="https://docs.joomla.org/Special:MyLanguage/Adding_a_new_article" class="btn btn-outline-secondary btn-lg px-4"><?php echo Text::_('COM_CONTENT_BLANKSTATE_BUTTON_LEARNMORE'); ?></a>
```

**Comment by brianteeman** [comment]
Same thing happens if I only have unpublished items. So there is a bug there.

**Comment by PhilETaylor** [comment]
> No that should never happen. It was specifically tested for during design.

https://user-images.githubusercontent.com/400092/115880426-9f291a80-a442-11eb-81a4-8636e38b5f5a.mp4

**Comment by PhilETaylor** [comment]
> Same thing happens if I only have unpublished items. So there is a bug there.

ok, well one I cannot replicate at the moment - see the video.

File: administrator/components/com_content/tmpl/articles/blankstate.php (hunk 1 above)
Comment (PhilETaylor) [reply administrator/components/com_content/tmpl/articles/blankstate.php]: To be honest I never expected this to be the final url...

**Comment by brianteeman** [comment]
Something weird is going on - trying to replicate it reliably

**Comment by Bond97** [comment]
I think, that it will be better to add this page "looks like..." to other menu items under the same visibility conditions: categories, featured articles, fields, field groups and menus. 

@PhilETaylor what do you think about it?

**Comment by brianteeman** [comment]
OK so if you set a filter and the filter returns no results you get the blank state which is clearly not what you intended.

issue

**Comment by PhilETaylor** [comment]
does that only happen with archived?

**Comment by PhilETaylor** [comment]
I cannot replicate that - I can select any state and never see the blank state - I just see no matching results.

**Comment by brianteeman** [comment]
issue

**Comment by PhilETaylor** [comment]
yeah I can replicate that now - when you have a category selected - ok, coffee and then I'll take a look. I was trying to implement the check in the db layer so that it can be generic and reusable.

**Comment by brianteeman** [comment]
Phil I can replicate it in any scenario where there are no matching results and I assume that you only meant this to be displayed when there are no records at all

**Comment by PhilETaylor** [comment]
> I think, that it will be better to add this page "looks like..." to other menu items under the same visibility conditions: categories, featured articles, fields, field groups and menus.

Read my opening comments. 

"This PR is a PROOF OF CONCEPT for Articles ONLY - if it received praise then I might do the others, and others can join in too.."

The "Others" meaning the other left menu items.

**Comment by PhilETaylor** [comment]
> Phil I can replicate it in any scenario where there are no matching results and I assume that you only meant this to be displayed when there are no records at all

Im looking, but that is certainly not what I get. The ONLY time you should ever see blank state is if the database table is is totally empty (has zero rows in the database as looked up by `SELECT count(*) from #__TABLENAME` )

**Comment by PhilETaylor** [comment]
Coffee fixes everything...

**Comment by PhilETaylor** [comment]
pull again now - should be fixed

**Comment by brianteeman** [comment]
Yes that did it. This is a great improvement.

File: administrator/components/com_content/tmpl/articles/blankstate.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+
+?>
+<form action="<?php echo Route::_('index.php?option=com_content&view=articles'); ?>" method="post" name="adminForm" id="adminForm">
+
+	<div class="px-4 py-5 my-5 text-center">
+		<span aria-hidden="true" class="fa-8x icon-copy mb-4 article"></span>
```
Comment (brianteeman) [inline administrator/components/com_content/tmpl/articles/blankstate.php]:
```suggestion
		<span class="fa-8x icon-copy mb-4 article" aria-hidden="true" ></span>
```

Just for consistency

File: administrator/language/en-GB/com_content.ini [hunk 3]
```diff
@@ -12,6 +12,10 @@ COM_CONTENT_ATTRIBS_ARTICLE_SETTINGS_LABEL="Options"
 COM_CONTENT_ATTRIBS_FIELDSET_LABEL="Options"
 COM_CONTENT_BATCH_OPTIONS="Batch process the selected articles"
 COM_CONTENT_BATCH_TIP="If a category is selected for move/copy, any actions selected will be applied to the copied or moved articles. Otherwise, all actions are applied to the selected articles."
+COM_CONTENT_BLANKSTATE_BUTTON_ADD="Add your first article"
+COM_CONTENT_BLANKSTATE_BUTTON_LEARNMORE="Learn more"
+COM_CONTENT_BLANKSTATE_CONTENT="Articles are the main content on most Joomla sites. You can use the Joomla Article Editor to create and manage your content."
+COM_CONTENT_BLANKSTATE_TITLE="Looks like you don't have any Articles yet"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_content.ini]:
```suggestion
COM_CONTENT_BLANKSTATE_TITLE="No Articles have been created yet."
```

The other strings are fine

**Comment by brianteeman** [comment]
Happy to work through other components and create PR for those. ITs easy enough now you did the hard work.

**Comment by Bakual** [comment]
I would actually use `$this->setLayout('blankstate')` in the view instead of requiring the blankstate layout in the default layout. This would make the code simpler and you could even get rid of `$this->isBlankState`.

**Comment by PhilETaylor** [comment]
The hard part is the reusable part in `libraries/src/MVC/Model/ListModel.php ` - the hard thing was to try and make it generic and reusable without having to code it for each component 

> I would actually use $this->setLayout('blankstate') 

DAMN... I tried everything to try to work that out.... let me try that... I was setting $tpl...

**Comment by Bakual** [comment]
Hehe, I use it in my component, that's why I knew it existed. I remember that I first tried with $tpl as well back then :-)

**Comment by PhilETaylor** [comment]
Done. Love it.

**Comment by brianteeman** [comment]
So just the language string and I can mark it as a successful test

**Comment by PhilETaylor** [comment]
> So just the language string and I can mark it as a successful test

I dont see any outstanding changes Brian, maybe GitHub caching ?

**Comment by PhilETaylor** [comment]
Screenshot now updated in initial comment.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on f28b2f78b71f845397822c0d45b6d701e99225e0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33264">issues.joomla.org/tracker/joomla-cms/33264</a>.</sub>

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on 55d5cde30cda923bf063ff088120e8be4731326d<br /><br />Fire!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33264">issues.joomla.org/tracker/joomla-cms/33264</a>.</sub>

**Comment by PhilETaylor** [comment]
Ive added the blank state for Featured Articles page too - just as it is still com_content and very quick to add to this PR. Sorry, should not really invalidate your tests, but a quick check might be appreciated.

File: libraries/src/MVC/Model/ListModel.php [hunk 4]
```diff
@@ -17,6 +17,7 @@
 use Joomla\CMS\MVC\Factory\MVCFactoryInterface;
 use Joomla\CMS\Pagination\Pagination;
 use Joomla\Database\DatabaseQuery;
+use Joomla\Database\Mysqli\MysqliQuery;
```
Comment (richard67) [inline libraries/src/MVC/Model/ListModel.php]: Is this right? What if using a PostgreSQL database?

File: libraries/src/MVC/Model/ListModel.php [hunk 5]
```diff
@@ -149,6 +150,28 @@ public function __construct($config = array(), MVCFactoryInterface $factory = nu
...
+	/**
+	 * Is this a blank state, I.e: no items of this type regardless of the searched for states.
+	 *
+	 * @return boolean
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	public function getisBlankSlate()
+	{
+		/** @var MysqliQuery $sql */
```
Comment (richard67) [inline libraries/src/MVC/Model/ListModel.php]: See previous review comment.

File: libraries/src/MVC/Model/ListModel.php (hunk 4 above)
Comment (PhilETaylor) [reply libraries/src/MVC/Model/ListModel.php]: Its 2021 and Ive never used PostgreSQL ever.. even when I worked at Skype.com and their main db was PostgreSQL haha...  Its just a comment, I can remove it.

Comment (PhilETaylor) [reply libraries/src/MVC/Model/ListModel.php]: Done. Its gone.

File: libraries/src/MVC/Model/ListModel.php (hunk 5 above)
Comment (PhilETaylor) [reply libraries/src/MVC/Model/ListModel.php]: Its Gone.

**Comment by richard67** [comment]
@PhilETaylor The system tests are failing constantly at the same place, I have restarted them 2 times already. They fail at "ContentListCest: Test that it loads without php notices and warnings.". Not sure if it's related to your PR, but it could be.

**Comment by richard67** [comment]
P.S.: https://ci.joomla.org/joomla/joomla-cms/42311/1/22

**Comment by PhilETaylor** [comment]
```
$I->wantToTest('that it loads without php notices and warnings.');
		$I->amOnPage(ContentListPage::$url);
		$I->waitForElement(ContentListPage::$filterSearch);
		$I->checkForPhpNoticesOrWarnings();
```

Ok Im guessing this test is looking for the filterSearch options stuff and obviously those no longer show on a blank state.  I'll try to fix the tests, as the tests are now wrong, not the new code.

**Comment by richard67** [comment]
> ... as the tests are now wrong, not the new code.

Exactly that was my assumption.

**Comment by PhilETaylor** [comment]
```
OK, but incomplete, skipped, or risky tests!
--
1471 | Tests: 41, Assertions: 214, Skipped: 4.
```

Tests look fixed.

**Comment by PhilETaylor** [comment]
I lied :(

**Comment by richard67** [comment]
@PhilETaylor MySQL (<8) and MySQL 8 system test passed. Only the PostgreSQL test was failing at that place with a timeout. Will restart drone to see if it was maybe just the unluck we have sometimes.

**Comment by PhilETaylor** [comment]
failed again. No idea. Is there a way to get the screenshot/html that it says it saved when the test failed?

**Comment by richard67** [comment]
@PhilETaylor Do you have access to this? https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33264/system-tests/42315/ ... the link can be found when choosing the "arifacts-system-tests" in the drone results.

**Comment by richard67** [comment]
ContentListCest loadsWithoutPhpNoticesAndWarnings postgres fail

**Comment by PhilETaylor** [comment]
> 42803 42803, 7, ERROR: column "a.id" must appear in the GROUP BY clause or be used in an aggregate function LINE 13: ORDER BY a.id desc LIMIT 20 ^

strange... as my query has no limit of 20 :) 

I guess Im spending the night learning postgres

**Comment by richard67** [comment]
SQL syntax. PostgreSQL is more strict with that, it seems. Do you understand it or shall I help later after my JBS meeting has finished?

**Comment by PhilETaylor** [comment]
I have replicated it locally now, now to debug, like I said, my only SQL change has no limit so must be something else...

**Comment by PhilETaylor** [comment]
fixed locally  :)

**Comment by richard67** [comment]
@PhilETaylor I guess that was it. But it's not only PostgreSQL which requires it, MySQL would require it too if sunning in strict mode, Oracle would require it (if we would support that).

**Comment by PhilETaylor** [comment]
Screenshot 2021-04-23 at 17 40 20

Be right back, off to put "Postgres expert" on my CV...

File: administrator/components/com_content/tmpl/articles/blankstate.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+
+?>
+<form action="<?php echo Route::_('index.php?option=com_content&view=articles'); ?>" method="post" name="adminForm" id="adminForm">
+
+	<div class="px-4 py-5 my-5 text-center">
+		<span class="fa-8x icon-copy mb-4 article" aria-hidden="true" ></span>
```
Comment (Quy) [inline administrator/components/com_content/tmpl/articles/blankstate.php]:
```suggestion
		<span class="fa-8x icon-copy mb-4 article" aria-hidden="true"></span>
```

**Comment by PhilETaylor** [comment]
Thanks @Quy

**Comment by wilsonge** [comment]
I love this :) Lets get some tests and get it in. Thankyou so much!!

**Comment by TLWebdesign** [comment]
I have tested this item :white_check_mark: successfully on 8d3accc2cb7b651cbfe71f0facacc66299f0fc9d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33264">issues.joomla.org/tracker/joomla-cms/33264</a>.</sub>

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on 8d3accc2cb7b651cbfe71f0facacc66299f0fc9d<br /><br />Very nice feature! Thanks!
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33264">issues.joomla.org/tracker/joomla-cms/33264</a>.</sub>

**Comment by richard67** [comment]
I've advertised a bit a JoomlaDay USA, so we did not have to wait very long for testers 😄

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33264">issues.joomla.org/tracker/joomla-cms/33264</a>.</sub>

**Comment by PhilETaylor** [comment]
> I've advertised a bit a JoomlaDay USA 

And to think, I turned down a personal invitation to be their major sponsor this year... haha.

**Pull request merged**